analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,214 @@
+                self.tag_IPs_b_LL = False
+                raise AttributeError('You specified a numerator without a denominator for the log likelihood based on IP significance')
+            else :
+                self.tag_IPs_b_LL = True
+                self.num_IPs_file = TFile.Open(self.cfg_ana.num_IPs[0])
+                self.num_IPs_hist = self.num_IPs_file.Get(self.cfg_ana.num_IPs[1])
+                self.denom_IPs_file = TFile.Open(self.cfg_ana.denom_IPs[0])
+                self.denom_IPs_hist = self.denom_IPs_file.Get(self.cfg_ana.denom_IPs[1])
+                self.ratio_IPs = TH1F("ratio_IPs","num_IPs over denom_IPs", self.num_IPs_hist.GetXaxis().GetNbins(), self.num_IPs_hist.GetXaxis().GetXmin(), self.num_IPs_hist.GetXaxis().GetXmax())
+                self.ratio_IPs.Divide(self.num_IPs_hist,self.denom_IPs_hist)
+
+    
     def process(self, event):
         assumed_vertex = TVector3(0, 0, 0)
         jets = getattr(event, self.cfg_ana.jets)
         for jet in jets:
-            #for ptc in jet.constituents[211]:
-            #    IPsn = 1
-            #    N_charge = 0
-                #for i in range(len(const)):
-                #    if const[i][1][0]._charge <> 0:
-                #        if len const[i][1] > 0:
-                #            for l in range(len(const[i][1]))
-                #                partic = const[i][1][l]
-                #                h = partic.path
-                #                Helix.compute_IP(h, assumed_vertex)
-                #                setattr(partic, self.IP, h.IP)
-                #                setattr(partic, self.tIP, h.closest.t)
-                #                setattr(partic, self.xIP, h.closest.x)
-                #                setattr(partic, self.yIP, h.closest.y)
-                #                setattr(partic, self.zIP, h.closest.z)
-                #                IPsn *= h.IP
-                #                N_charge += 1
-                #        else :
-                #            pass
-                #    else :
-                #        pass
-                #if N_charge <> 0 :
-                #    IPs = IPSn**(1.0/N_charge)
-                #else :
-                #    pass
+            IP_b_LL = 0     # value of the log likelihood ratio based on IP initiated at 0
+            IPs_b_LL = 0    # value of the log likelihood ratio based on IP significance initiated at 0
+            ipsig_ptcs = [] # list of IP signif and associated ptcs
             for id, ptcs in jet.constituents.iteritems():
-                if id in [22,130]:
+                if abs(id) in [22,130,11]:
                     continue
                 for ptc in ptcs :
-                    if ptc._charge == 0 :
-                        continue
-                    ptc.path.compute_IP(assumed_vertex)
-                    #jet.IP = ptc.path.IP
-            #setattr(event, self.cfg_ana.output, IPs)
+                    if hasattr(ptc.path, "charge") :
```

Comment:
instead use the charge of the particle. 
you can stop here with pdb, and do dir(ptc) or help(ptc) to see which methods are available. 
do not test for existence of the charge method.